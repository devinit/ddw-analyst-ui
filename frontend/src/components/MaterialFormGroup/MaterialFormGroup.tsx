import * as React from 'react';
import classNames from 'classnames';
import { Col, Form, FormGroup, Row } from 'react-bootstrap';

interface MaterialFormGroupProps {
  touched?: boolean;
  focused?: boolean;
  errors?: string;
  onFocus?: (event: React.FocusEvent<HTMLInputElement>) => void;
  onBlur?: (event: React.FocusEvent<HTMLInputElement>) => void;
  onChange?: (event: React.ChangeEvent) => void;
  value?: string;
  md?: string | boolean | number;
  as?: string;
  id?: string;
  name?: string;
  required: boolean;
  label?: string;
}

export const MaterialFormGroup: React.SFC<MaterialFormGroupProps> = (props) => {
  const getFormGroupClasses = (value?: string): string => {
    return classNames('bmd-form-group', {
      'is-focused': props.focused,
      'is-filled': !!value,
    });
  };

  return (
    <Row>
      <Col md={props.md as any}>
        <FormGroup className={getFormGroupClasses(props.value)} data-testid="material-form-group">
          <Form.Label htmlFor={props.id || props.name} className="bmd-label-floating">
            {props.label}
          </Form.Label>
          <Form.Control
            as={props.as as any}
            id={props.id}
            required={props.required}
            name={props.name}
            onChange={props.onChange}
            isInvalid={!!props.errors}
            onFocus={props.onFocus}
            onBlur={props.onBlur}
          />
          <Form.Control.Feedback type="invalid" data-testid="material-form-control-feedback">
            {props.touched && props.errors ? props.errors : null}
          </Form.Control.Feedback>
        </FormGroup>
      </Col>
    </Row>
  );
};

export { MaterialFormGroup as default };
