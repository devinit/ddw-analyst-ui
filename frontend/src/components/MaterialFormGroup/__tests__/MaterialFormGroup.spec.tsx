/**
 * @jest-environment jsdom
 */
import '@testing-library/jest-dom/extend-expect';
import * as React from 'react';
import { render } from '@testing-library/react';
import * as TestRenderer from 'react-test-renderer';
import { MaterialFormGroup } from '../MaterialFormGroup';

test('renders the component correctly', () => {
  const renderer = TestRenderer.create(
    <MaterialFormGroup
      md="8"
      id="title"
      name="title"
      required
      onChange={ jest.fn() }
      onFocus={ jest.fn() }
      onBlur={ jest.fn() }
    />
  ).toJSON();

  expect(renderer).toMatchSnapshot();
});

test('renders the form group with the class is-focused when focused', () => {
  const { getByTestId } = render(
    <MaterialFormGroup
      md="8"
      id="title"
      name="title"
      required
      onChange={ jest.fn() }
      onFocus={ jest.fn() }
      onBlur={ jest.fn() }
      focused
    />
  );

  expect(getByTestId('material-form-group')).toHaveClass('is-focused');
});

test('renders the form group with the class is-filled when the input has a value', () => {
  const { getByTestId } = render(
    <MaterialFormGroup
      md="8"
      id="title"
      name="title"
      required
      onChange={ jest.fn() }
      onFocus={ jest.fn() }
      onBlur={ jest.fn() }
      value="testing"
    />
  );

  expect(getByTestId('material-form-group')).toHaveClass('is-filled');
});

test('does not show errors when not touched', () => {
  const { getByTestId } = render(
    <MaterialFormGroup
      md="8"
      id="title"
      name="title"
      required
      onChange={ jest.fn() }
      onFocus={ jest.fn() }
      onBlur={ jest.fn() }
      touched={ false }
      errors="This field is invalid"
    />
  );

  expect(getByTestId('material-form-control-feedback')).not.toHaveTextContent('This field is invalid');
});

test('shows supplied errors when touched', () => {
  const { getByTestId } = render(
    <MaterialFormGroup
      md="8"
      id="title"
      name="title"
      required
      onChange={ jest.fn() }
      onFocus={ jest.fn() }
      onBlur={ jest.fn() }
      touched
      errors="This field is invalid"
    />
  );

  expect(getByTestId('material-form-control-feedback')).toHaveTextContent('This field is invalid');
});
