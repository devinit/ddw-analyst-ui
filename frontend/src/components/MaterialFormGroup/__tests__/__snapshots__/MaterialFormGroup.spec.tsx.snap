// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the component correctly 1`] = `
<div
  className="row"
>
  <div
    className="col-md-8"
  >
    <div
      className="bmd-form-group form-group"
      data-testid="material-form-group"
    >
      <label
        className="bmd-label-floating form-label"
        htmlFor="title"
      />
      <input
        className="form-control"
        id="title"
        name="title"
        onBlur={[MockFunction]}
        onChange={[MockFunction]}
        onFocus={[MockFunction]}
        required={true}
      />
      <div
        className="invalid-feedback"
        data-testid="material-form-control-feedback"
      />
    </div>
  </div>
</div>
`;
